### PR TITLE
Make the DatasetPipeline repr more clear about optimized or unoptimiz…

### PR DESCRIPTION
**Pull request by jianoaix**
…ed num_stages

Signed-off-by: jianoaix <iamjianxiao@gmail.com>

## Why are these changes needed?
Deconfuse the repr message, based on user feedback

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset_pipeline.py [hunk 1]
```diff
@@ -1123,8 +1123,15 @@ def from_iterable(
         return DatasetPipeline(iterable, False, length=length)
 
     def __repr__(self) -> str:
-        return "DatasetPipeline(num_windows={}, num_stages={})".format(
-            self._length, 1 + len(self._stages)
+        if self._executed[0]:
+            return (
+                f"DatasetPipeline(num_windows={self._length}, "
+                f"num_stages_unoptimized={len(self._stages)}, "
+                f"num_stages_optimized={len(self._optimized_stages)})"
```
Comment (matthewdeng) [inline python/ray/data/dataset_pipeline.py:1130]: What is an optimized/unoptimized stage?

Comment (c21) [reply python/ray/data/dataset_pipeline.py:1130]: AFAIK optimized stage is stage by fusing multiple original stages together. btw why we need to expose the information here? curious what's the use case?

Comment (matthewdeng) [reply python/ray/data/dataset_pipeline.py:1130]: Yeah my confusion here is about how to interpret this as a generic user. Some thoughts:

1. It's not clear what is counted as an unoptimized stage (i.e. is it the original number of stages or the stages that are yet to be optimized).
2. This depends on whether or not stages have been executed to trigger optimization.
3. Should this information be available in `Dataset`s as well?

Comment (ericl) [reply python/ray/data/dataset_pipeline.py:1130]: How about just calling this `num_stages` and `num_stages_fused`? I don't think we use the term optimize but do say fusion.

Comment (ericl) [reply python/ray/data/dataset_pipeline.py:1130]: Hmm seems the strange part is that we don't show the stages after optimization in most cases (user prints pipeline prior to execution).

Is it possible to run optimization before generating the repr? That hopefully shouldn't be a big overhead.

Comment (jianoaix) [reply python/ray/data/dataset_pipeline.py:1130]: I'm just studying this magic method and `__str__' in Python, what I see the convention is:
- `__repr__` is for developers (who develop the lib): it should capture what's deciding this class ("unambiguous"). In this case, it should be base_datasets and stages. Similar for Dataset, it's input_blocks and plan.
- `__str__` is for users (who use the lib): it should make it more readable and user friendly, such as schema, num of windows (I think stages is optional to mention). For Dataset, it's current `__repr__` which print schema and num of rows etc (which are user oriented).

Comment (jianoaix) [reply python/ray/data/dataset_pipeline.py]: Narrow down the change to just fix the existing num_stages. PTAL.

File: python/ray/data/dataset_pipeline.py [hunk 2]
```diff
@@ -1123,8 +1123,9 @@ def from_iterable(
         return DatasetPipeline(iterable, False, length=length)
 
     def __repr__(self) -> str:
-        return "DatasetPipeline(num_windows={}, num_stages={})".format(
-            self._length, 1 + len(self._stages)
+        return (
+            f"DatasetPipeline(num_windows={self._length}, "
+            f"num_stages={len(self._stages)})"
```
Comment (c21) [inline python/ray/data/dataset_pipeline.py:1128]: dumb question - why we have `1 +` before? is it for the read stage?

Comment (jianoaix) [reply python/ray/data/dataset_pipeline.py:1128]: I assume we might have treated "window" or "repeat" as a stage.

**Comment by jianoaix** [comment]
This confused users again recently: https://discuss.ray.io/t/just-two-stages-present-no-matter-how-many-stages-defined-for-datasetpipeline/8056/5

**Pull request closed without merging**